### PR TITLE
Fixes memory leak caused by error string.

### LINKED ISSUE #8: CI fails in ubuntu often, I suspect a segfault, have no idea how to diagnose
See the toils and travails in #6 -- _many_ failures, some of those commits definitely deserved to fail, but even after I cleaned up the code considerably, I see "segfaults". Not sure if it's the C library doing something untoward, but it happens extremely often in Ubuntu and never in macOS (I develop in macOS, and the tests work a frustrating 99.999% of the time, though I've managed to see errors too.) I just don't know enough C, Haskell or CI to get to the bottom of this, and I spent an inordinate amount of hours the past two nights staring at this.

### PR DESCRIPTION
**Pull request by lfborjas**
Closes #8 

Turns out allocating a pointer for the error string, vs. the `char [256]` that the [swiss ephemeris documentation](https://www.astro.com/swisseph/swephprg.htm#_Toc49847825) mentions repeatedly was causing a memory leak. We now allocate space for only 256 characters -- no more segfaults!

Also, some "undefined behavior" is also warned against in the documentation for [`Foreign.Marshal.Alloc`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Foreign-Marshal-Alloc.html#g:2). After reading more, I guess I _could_ use [`allocaBytes`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Foreign-Marshal-Alloc.html#v:allocaBytes) but I'm not 100% confident that a `char` is 1 byte in all platforms? I'd rather have Haskell know that it must allocate an array large enough for 256 [`CChars`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Foreign-C-Types.html#t:CChar). Another option is the more explicit [`CStringLen`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Foreign-C-String.html#t:CStringLen) type, though I'm not sure how much clearer that would be than hewing close to the original library and going with an array of 256 chars explicitly?

You can see in the code itself that `serr` is usually defined as a char array of size `AS_MAXCH` which is defined in:

https://github.com/lfborjas/swiss-ephemeris/blob/c31d31286e708537e89b857dc7f607ea8eb3b48d/csrc/sweodef.h#L266

Most conspicuously, in `swe_calc`, which is called by `swe_calc_ut`:

https://github.com/lfborjas/swiss-ephemeris/blob/c31d31286e708537e89b857dc7f607ea8eb3b48d/csrc/sweph.c#L581

tl;dr: RTFM, the signature was there all along:

```
long swe_calc_ut(

double tjd_ut,       /* Julian day number, Universal Time */

int ipl,                  /* planet number */

long iflag,          /* flag bits */

double *xx,          /* target address for 6 position values: longitude, latitude, distance, long. speed, lat. speed, dist. speed */

char *serr);         /* 256 bytes for error string */
```

## Notes

Some good references for tooling:

* https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes
* https://markkarpov.com/post/github-actions-for-haskell-ci.html

And for actual troubleshooting the potential segfault:

https://github.com/ndmitchell/spaceleak

In the end, I was able to reproduce the issue more reliably in my mac mini by building the tests with 

    stack --profile test

And running them with the `+RTS -K1K -s` options as per the above link -- one should be able to send these to `stack --profile test`, but I wasn't able to figure it out, this proved to be reliable however:

    .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/build/swiss-ephemeris-test/swiss-ephemeris-test +RTS -K1K -s

Which also prints some interesting memory usage information:

```
      19,366,312 bytes allocated in the heap
         492,760 bytes copied during GC
         260,944 bytes maximum residency (2 sample(s))
          76,720 bytes maximum slop
               0 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0        15 colls,    15 par    0.006s   0.002s     0.0001s    0.0006s
  Gen  1         2 colls,     1 par    0.001s   0.001s     0.0003s    0.0004s

  Parallel GC work balance: 49.74% (serial 0%, perfect 100%)

  TASKS: 20 (1 bound, 19 peak workers (19 total), using -N6)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.009s elapsed)
  MUT     time    0.068s  (  0.080s elapsed)
  GC      time    0.007s  (  0.002s elapsed)
  RP      time    0.000s  (  0.000s elapsed)
  PROF    time    0.000s  (  0.000s elapsed)
  EXIT    time    0.000s  (  0.001s elapsed)
  Total   time    0.076s  (  0.092s elapsed)

  Alloc rate    283,718,073 bytes per MUT second

  Productivity  89.3% of total user, 87.4% of total elapsed
```

When the memory leak was still present, I managed to get it from the above command, and also from some `gdb` sessions:

```
luis@mac-mini ~/c/l/swiss-ephemeris (late-night-ci-poking)> gdb --args .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/build/swiss-ephemeris-test/swiss-ephemeris-test +RTS -K1K -V0
GNU gdb (GDB) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin19.5.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/build/swiss-ephemeris-test/swiss-ephemeris-test...
Reading symbols from /Users/luis/code/lfborjas/swiss-ephemeris/.stack-work/dist/x86_64-osx/Cabal-3.0.1.0/build/swiss-ephemeris-test/swiss-ephemeris-test.dSYM/Contents/Resources/DWARF/swiss-ephemeris-test...
(gdb) run
Starting program: /Users/luis/code/lfborjas/swiss-ephemeris/.stack-work/dist/x86_64-osx/Cabal-3.0.1.0/build/swiss-ephemeris-test/swiss-ephemeris-test +RTS -K1K -V0
[New Thread 0x2803 of process 84588]
[New Thread 0x1b03 of process 84588]
[New Thread 0x2603 of process 84588]
warning: unhandled dyld version (16)

SwissEphemeris
  calculateCoordinates
    calculates coordinates for the Sun for a specific day
    fails to calculate coordinates for Chiron if no ephemeris file is set
  calculateCuspsStrict
    calculates cusps and angles for a given place and time, keeping the same house system (not near the poles)
    fails when using a house system that is unable to calculate cusps near the poles
  calculateCuspsLenient
    falls back to Porphyry when calculating cusps for a place near the poles
[New Thread 0x1a0f of process 84588]
[New Thread 0x1c03 of process 84588]
[New Thread 0x1d03 of process 84588]
[New Thread 0x1e03 of process 84588]
[New Thread 0x1f03 of process 84588]
[New Thread 0x2003 of process 84588]
[New Thread 0x2103 of process 84588]
[New Thread 0x2203 of process 84588]
[New Thread 0x2303 of process 84588]
[New Thread 0x2403 of process 84588]
[New Thread 0x2503 of process 84588]
[New Thread 0x2a03 of process 84588]
[New Thread 0x2b03 of process 84588]
[New Thread 0x2c03 of process 84588]
[New Thread 0x2d03 of process 84588]
[New Thread 0x2e03 of process 84588]
[New Thread 0x5203 of process 84588]
[New Thread 0x5303 of process 84588]
[New Thread 0x5403 of process 84588]

Thread 4 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x1a0f of process 84588]
evacuate (p=<optimized out>) at /Users/builder/Desktop/builds/V164sjj5/0/ghc/ghc/rts/sm/Evac.c:521
521	/Users/builder/Desktop/builds/V164sjj5/0/ghc/ghc/rts/sm/Evac.c: No such file or directory.
(gdb) bt
#0  evacuate (p=<optimized out>) at /Users/builder/Desktop/builds/V164sjj5/0/ghc/ghc/rts/sm/Evac.c:521
#1  0x0000004200626628 in ?? ()
#2  0x0000004200626628 in ?? ()
#3  0x0000000000000000 in ?? ()
(gdb) 
```

... which shows that things went very wrong very early on. Also that apparently `hspec` runs threaded?

### CODE REVIEW

**Comment by lfborjas** [comment]
Browsing the mailing list for swiss ephemeris, I found [this message](https://groups.io/g/swisseph/message/8679) (reproduced below), which pinpoints _exactly_ the problem I was having. The issue hasn't been fixed in 2.09.03, unfortunately, but the workaround they propose (to manually initialize the error string,) is exactly what I ended up doing! 

```
Thank you for making the release of Swiss Ephemeris version 2.09.02 available. :) I compiled the sources from https://www.astro.com/ftp/swisseph/src/ without any warnings. Unfortunately, I am now getting a crash in swecalc() when computing asteroids (sweph.c line 1070, soon after the "do_asteroid:" label) which happens after the following strcpy:
 
if (serr != NULL) { strcpy(serr2, serr); *serr = '\0'; }
 
The crash happens because serr is uninitialized, so long garbage is copied into finite buffer serr2. The buffer serr used to be initialized to the empty string in swecalc() (sweph.c line 583) however that initialization line has been commented out in the newest source code, as seen below:
 
//if (serr != NULL) *serr = '\0';  // is done in calling function
 
Sure enough, in calling function swe_calc(), the new line below has been added (sweph.c line 347). However, that new line is within an #ifdef TRACE block. Most Swiss Ephemeris clients don't have TRACE defined, so *serr doesn't get initialized, and the crash above happens.
 
if (serr != NULL) *serr = '\0';
 
To fix this, just move the *serr initialization to the very top of swe_calc() right before "#ifdef TRACE" so it always runs. A client can also workaround by manually initializing *serr before calling swe_calc() (although that was never never necessary before).
 
```

**Comment by lfborjas** [comment]
For posterity: what's described above, at least in my architecture, can be reproduced thusly:

```c
#include <stdio.h>
#include <string.h>

int
main(void){
        char* serr ;
        // this is equivalent to something that swecalc does:
        char serr2[256];
        *serr2 = '\0';
        strcpy(serr2, serr);
        printf(serr2);
        return 0;
}
```

When run:
```
gcc -Wall oops5.c -o oops5
oops5.c:16:9: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
        printf(serr2);
               ^~~~~
oops5.c:16:9: note: treat the string as an argument to avoid this
        printf(serr2);
               ^
               "%s", 
1 warning generated.
> ./oops5
�|�s�⏎                                                        
```

However, simply giving `serr` a size seems to fix things:

```c
#include <stdio.h>
#include <string.h>

int
main(void){
        char serr[256] ;
        //*serr = '\0';
        // this is equivalent to something that swecalc does:
        char serr2[256];
        *serr2 = '\0';
        strcpy(serr2, serr);
        printf(serr2);
        return 0;
}
```

```
luis@mac-mini ~/c/l/swiss-ephemeris (fix-tests-memory-patch)> gcc -Wall oops5.c -o oops5
oops5.c:16:9: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
        printf(serr2);
               ^~~~~
oops5.c:16:9: note: treat the string as an argument to avoid this
        printf(serr2);
               ^
               "%s", 
1 warning generated.
> ./oops5 
# prints nothing, as it should.
>
```
